### PR TITLE
Fix for IDENTITY-2925

### PR DESCRIPTION
**Pull request by malithie**
Attached fix provides support to add custom grant types by allowing to add grant type validators. Previously only default validators were supported which blocked any other grant type that pass by.

### CODE REVIEW

File: components/identity/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/OAuthRequestWrapper.java [hunk 1]
```diff
@@ -21,13 +21,26 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletRequestWrapper;
 import javax.ws.rs.core.MultivaluedMap;
+import java.util.*;
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/OAuthRequestWrapper.java]: Remove imports with *. Each class should be imported individually

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/cache/AppInfoCache.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/*
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/cache/AppInfoCache.java]: Use correct license headers . You can refer to https://github.com/wso2-dev/carbon-identity/pull/33/files

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/cache/AppInfoCache.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.oauth.cache;
+
+import org.wso2.carbon.identity.core.model.OAuthAppDO;
+import org.wso2.carbon.utils.CarbonUtils;
+
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/cache/AppInfoCache.java]: Add class level doc comments and add comments for at least public methods

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 4]
```diff
@@ -39,6 +45,7 @@
 import org.wso2.carbon.identity.openidconnect.IDTokenBuilder;
 import org.wso2.carbon.utils.CarbonUtils;
 
+import javax.servlet.http.HttpServletRequest;
 import javax.xml.namespace.QName;
 import java.util.*;
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Remove imports with *

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 5]
```diff
@@ -374,6 +387,56 @@ public boolean isRefreshTokenRenewalEnabled(){
...
+                            try {
+                                @SuppressWarnings("unchecked")
+                                Class<? extends OAuthValidator<HttpServletRequest>>
+                                        oauthValidatorClass =
+                                        (Class<? extends OAuthValidator<HttpServletRequest>>) Class
+                                                .forName(entry.getValue());
+                                supportedGrantTypeValidators
+                                        .put(entry.getKey(), oauthValidatorClass);
+                            } catch (ClassNotFoundException e) {
+                                log.error("Cannot find class: " + entry.getValue());
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Log the exception as well.

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 6]
```diff
@@ -374,6 +387,56 @@ public boolean isRefreshTokenRenewalEnabled(){
...
+                            } catch (ClassCastException e) {
+                                log.error("Cannot cast class: " + entry.getValue());
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Log the exception

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 7]
```diff
@@ -896,56 +959,85 @@ private void parseTokenPersistenceProcessorConfig(OMElement oauthConfigElem) {
...
+                                    getQNameWithIdentityNS(
+                                            ConfigElements.GRANT_TYPE_VALIDATOR_IMPL_CLASS));
+
+                    String authzGrantValidatorImplClass = null;
+                    if (authzGrantValidatorClassNameElement != null) {
+                        authzGrantValidatorImplClass =
+                                authzGrantValidatorClassNameElement.getText();
+                    }
+
+                    if (authzGrantValidatorImplClass != null &&
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Use String utils to check Strings.

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 8]
```diff
@@ -896,56 +959,85 @@ private void parseTokenPersistenceProcessorConfig(OMElement oauthConfigElem) {
...
-            defaultGrantTypes.put(org.wso2.carbon.identity.oauth.common.GrantType.SAML20_BEARER.toString(), "org.wso2.carbon.identity.oauth2.token.handlers.grant.saml.SAML2BearerGrantHandler");
-            defaultGrantTypes.put(org.wso2.carbon.identity.oauth.common.GrantType.IWA_NTLM.toString(),"org.wso2.carbon.identity.oauth2.token.handlers.grant.iwa.ntlm.NTLMAuthenticationGrantHandler");
+                     "Therefore instantiating default grant type handlers");
+
+            Map<String, String> defaultGrantTypes = new Hashtable<String, String>(5);
+            defaultGrantTypes.put(GrantType.AUTHORIZATION_CODE.toString(),
+                                  "org.wso2.carbon.identity.oauth2.token.handlers.grant.AuthorizationCodeHandler");
+            defaultGrantTypes.put(GrantType.CLIENT_CREDENTIALS.toString(),
+                                  "org.wso2.carbon.identity.oauth2.token.handlers.grant.ClientCredentialsGrantHandler");
+            defaultGrantTypes.put(GrantType.PASSWORD.toString(),
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Cant we define these class names as constants ?

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthTokenRequest.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2005-2013, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthTokenRequest.java]: Add correct license headers

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthTokenRequest.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    private String windows_token;
+
+    private String tenantDomain;
+
+    private RequestParameter[] requestParameters;
+
+    public CarbonOAuthTokenRequest(HttpServletRequest request) throws OAuthSystemException,
+                                                                      OAuthProblemException {
+        super(request);
+        assertion = request.getParameter("assertion");
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthTokenRequest.java]: Can we define these as constants ?

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthTokenRequest.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+import org.apache.amber.oauth2.common.utils.OAuthUtils;
+import org.apache.amber.oauth2.common.validators.OAuthValidator;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.oauth.config.OAuthServerConfiguration;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthTokenRequest.java]: Add class level doc comments and comments for at least public methods

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthTokenRequest.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+public class CarbonOAuthTokenRequest extends OAuthTokenRequest {
+    private static Log log = LogFactory.getLog(CarbonOAuthTokenRequest.class);
+
+    private String assertion;
+
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthTokenRequest.java]: Remove unnecessary new lines

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/RequestParameter.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2005-2013, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/RequestParameter.java]: Correct license headers

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java [hunk 14]
```diff
@@ -331,4 +335,39 @@ public boolean validateScope(OAuthTokenReqMessageContext tokReqMsgCtx)
...
+                !oAuthAppDO.getGrantTypes().contains(grantType)) {
+                if (log.isDebugEnabled()) {
+                    //Do not change this log format as these logs use by external applications
+                    log.debug("Unsupported Grant Type : " + grantType +
+                              " for client id : " + tokenReqDTO.getClientId());
+                }
+                return false;
+            }
+        } catch (InvalidOAuthClientException e) {
+            log.error("Error while reading application data ", e);
```
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java]: Can we add more context information here ?

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 5 above)
Comment (hasinthaindrajee) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Add more context data if available

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/cache/AppInfoCache.java (hunk 2 above)
Comment (malithie) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/cache/AppInfoCache.java]: Fixed with fce835d

File: components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/cache/AppInfoCache.java (hunk 3 above)
Comment (malithie) [inline components/identity/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/cache/AppInfoCache.java]: Fixed with fce835d

**Pull request merged**